src/main/java/com/hazelcast/gcp/GcpClient.java

```diff
@@ -58,34 +63,59 @@
             return gcpConfig.getProjects();
         }
         LOGGER.finest("Property 'projects' not configured, fetching the current GCP project");
+
         return singletonList(RetryUtils.retry(new Callable<String>() {
             @Override
             public String call() {
-                return gcpMetadataApi.currentProject();
+                String projectFromMetadataApi = gcpMetadataApi.currentProject();
+                // Check if valid project was retrieved
+                if (projectFromMetadataApi.startsWith("<!DOCTYPE html")) {
+                    LOGGER.severe("Project name could not be retrieved. Please specify the project name in the "
+                            + "configuration if running from outside the GCP network where your hazelcast cluster is.");
+                    throw new HazelcastException("Project could not be retrieved from GCP config");
+                }
+
+                return projectFromMetadataApi;
             }
-        }, RETRIES));
+        }, RETRIES, NON_RETRYABLE_KEYWORDS));
     }
 
     private List<String> zonesFromConfigOrMetadataApi(final GcpConfig gcpConfig) {
         if (!gcpConfig.getZones().isEmpty()) {
             return gcpConfig.getZones();
         }
         LOGGER.finest("Property 'zones' not configured, fetching the current GCP zone");
+
         return singletonList(RetryUtils.retry(new Callable<String>() {
             @Override
             public String call() {
-                return gcpMetadataApi.currentZone();
+                String currentZone = gcpMetadataApi.currentZone();
+                // Check if valid zone was retrieved
+                if ("html>".equals(currentZone)) {
```

Comment:
1. The same, please extract validation to a separate method.

2. Message can be something like `Zone could not be retrieved. Please specify the `zones` property if running from outside GCP network`